src/laddie/laddie_velocity.f90

```diff
@@ -0,0 +1,420 @@
+            laddie%viscV( ti) = laddie%viscV( ti) + (V_tot( tj) - V_tot( ti)) * Ah * npxref%H_c( ei) / mesh%TriA( ti) * mesh%TriCw( ti, ci) / D
+          END IF
+        END DO
+
+      END IF !(laddie%mask_b( ti)
+    END DO !ti = mesh%ti1, mesh%ti2
+      
+    ! Finalise routine path
+    CALL finalise_routine( routine_name)
+    
+  END SUBROUTINE compute_viscUV
+
+  SUBROUTINE compute_divQUV_upstream( mesh, laddie, npxref, Hstar_b)
+    ! Upstream scheme
+
+    IMPLICIT NONE
+
+    ! In/output variables:
+    TYPE(type_mesh),                        INTENT(IN)    :: mesh
+    TYPE(type_laddie_model),                INTENT(INOUT) :: laddie
+    TYPE(type_laddie_timestep),             INTENT(IN)    :: npxref
+    REAL(dp), DIMENSION(mesh%ti1:mesh%ti2), INTENT(IN)    :: Hstar_b
+
+    ! Local variables:
+    CHARACTER(LEN=256), PARAMETER                         :: routine_name = 'compute_divQUV_upstream'
+    REAL(dp), DIMENSION(mesh%nTri)                        :: U_tot, V_tot, H_b_tot
+    REAL(dp), DIMENSION(mesh%nE)                          :: U_c_tot, V_c_tot
+    INTEGER                                               :: ti, tj, ci, ei
+    REAL(dp)                                              :: D_x, D_y, D_c, u_perp_x, u_perp_y
+    LOGICAL, DIMENSION(mesh%nTri)                         :: mask_gl_b_tot, mask_cf_b_tot, mask_b_tot
+
+    ! Add routine to path
+    CALL init_routine( routine_name)
+
+    ! Calculate vertically averaged ice velocities on the edges
+    CALL gather_to_all_logical_1D( laddie%mask_gl_b, mask_gl_b_tot)
+    CALL gather_to_all_logical_1D( laddie%mask_cf_b, mask_cf_b_tot)
+    CALL gather_to_all_logical_1D( laddie%mask_b, mask_b_tot)
+    CALL gather_to_all_dp_1D( npxref%U, U_tot)
+    CALL gather_to_all_dp_1D( npxref%V, V_tot)
+    CALL gather_to_all_dp_1D( npxref%U_c, U_c_tot)
+    CALL gather_to_all_dp_1D( npxref%V_c, V_c_tot)
+    CALL gather_to_all_dp_1D( Hstar_b, H_b_tot)
+
+    ! Initialise with zeros
+    laddie%divQU = 0.0_dp
+    laddie%divQV = 0.0_dp
+
+    ! == Loop over triangles ==
+    ! =========================
+
+    DO ti = mesh%ti1, mesh%ti2
+
+      IF (laddie%mask_b( ti)) THEN
+
+        ! Loop over all connections of triangle ti
+        DO ci = 1, 3
+
+          tj = mesh%TriC( ti, ci)
+          ei = mesh%TriE( ti, ci)
+
+          !IF (par%master) THEN
+          !  WRITE( *, *) ti, ci, tj, ei, mask_gl_b_tot( tj)
+          !END IF     
+
+          ! Skip if no connecting triangle on this side
+          IF (tj == 0) CYCLE
+
+          ! Skip connection if neighbour is grounded. No flux across grounding line
+          IF (mask_gl_b_tot( tj)) CYCLE
+
+          ! The triangle-triangle vector from ti to tj
+          D_x = mesh%Tricc( tj,1) - mesh%Tricc( ti,1)
+          D_y = mesh%Tricc( tj,2) - mesh%Tricc( ti,2)
+          D_c = SQRT( D_x**2 + D_y**2)
+
+          ! Calculate vertically averaged ice velocity component perpendicular to this edge
```

Comment:
I think this is vertically averaged water velocity?

Reply:
Jup, will modify